napari/_qt/dialogs/qt_package_installer.py

```diff
@@ -0,0 +1,488 @@
+"""
+A tool-agnostic installation logic for the plugin manager.
+
+The main object is `InstallerQueue`, a `QProcess` subclass
+with the notion of a job queue. The queued jobs are represented
+by a `deque` of `*InstallerTool` dataclasses that contain the
+executable path, arguments and environment modifications.
+Available actions for each tool are `install`, `uninstall`
+and `cancel`.
+"""
+import atexit
+import contextlib
+import os
+import sys
+from collections import deque
+from dataclasses import dataclass
+from functools import lru_cache
+from logging import getLogger
+from pathlib import Path
+from tempfile import gettempdir, mkstemp
+from typing import Deque, Optional, Sequence, Tuple
+
+from npe2 import PluginManager
+from qtpy.QtCore import QObject, QProcess, QProcessEnvironment, Signal
+from qtpy.QtWidgets import QTextEdit
+
+from napari._version import version as _napari_version
+from napari._version import version_tuple as _napari_version_tuple
+from napari.plugins import plugin_manager
+from napari.plugins.pypi import _user_agent
+from napari.utils._appdirs import user_plugin_dir, user_site_packages
+from napari.utils.misc import StringEnum, running_as_bundled_app
+from napari.utils.translations import trans
+
+JobId = int
+log = getLogger(__name__)
+
+
+class InstallerActions(StringEnum):
+    "Available actions for the plugin manager"
+    install = "install"
+    uninstall = "uninstall"
+    cancel = "cancel"
+
+
+class InstallerTools(StringEnum):
+    "Available tools for InstallerQueue jobs"
+    conda = "conda"
+    pip = "pip"
+
+
+@dataclass(frozen=True)
+class AbstractInstallerTool:
+    action: InstallerActions
+    pkgs: Tuple[str, ...]
+    origins: Tuple[str, ...] = ()
+    prefix: Optional[str] = None
+
+    @property
+    def ident(self):
+        return hash((self.action, *self.pkgs, *self.origins, self.prefix))
+
+    # abstract method
+    @classmethod
+    def executable(cls):
+        "Path to the executable that will run the task"
+        raise NotImplementedError()
+
+    # abstract method
+    def arguments(self):
+        "Arguments supplied to the executable"
+        raise NotImplementedError()
+
+    # abstract method
+    def environment(
+        self, env: QProcessEnvironment = None
+    ) -> QProcessEnvironment:
+        "Changes needed in the environment variables."
+        raise NotImplementedError()
+
+    @staticmethod
+    def constraints() -> Sequence[str]:
```

Comment:
@jaimergp @goanpeca Am i reading right that this will close:
https://github.com/napari/napari/issues/5279
the issue that plugin installation can downgrade napari?
❤️

Reply:
YES IT DOES! 🎉 
```
The conflict is caused by:
    napari-n2v 0.0.4 depends on napari<=0.4.15
    napari-n2v 0.0.3 depends on napari<=0.4.15
    napari-n2v 0.0.2 depends on napari<=0.4.15
    napari-n2v 0.0.1 depends on napari<=0.4.15
    The user requested (constraint) napari==0.4.17rc4.dev86+g10035116

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

```